pkg/assets/assets.go

```diff
@@ -213,12 +216,165 @@ func ProcessTarGz(r io.Reader) (string, io.Reader, error) {
+	}
+
+	generic := make([]fmt.Stringer, 0)
+	for f := range zipFiles {
+		generic = append(generic, options.LiteralStringer(f))
+	}
+	selectedFile := options.Select("Select file to download:", generic).(fmt.Stringer).String()
+
+	fr := bytes.NewReader(zipFiles[selectedFile])
+
+	// return base of selected file since tar
+	// files usually have folders inside
+	return filepath.Base(selectedFile), fr, nil
+}
+
+func ProcessTarXz(r io.Reader) (string, io.Reader, error) {
+
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return "", nil, err
+	}
+	br := bytes.NewReader(b)
+
+	xr, err := xz.NewReader(br, 0)
+	if err != nil {
+		return "", nil, err
+	}
+
+	tr := tar.NewReader(xr)
+	tarFiles := []string{}
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return "", nil, err
+		}
+
+		if header.Typeflag == tar.TypeReg {
+			tarFiles = append(tarFiles, header.Name)
+		}
+	}
+	if len(tarFiles) == 0 {
+		return "", nil, errors.New("No files found in tar archive")
+	}
+
+	generic := make([]fmt.Stringer, 0)
+	for _, f := range tarFiles {
+		generic = append(generic, options.LiteralStringer(f))
+	}
+	selectedFile := options.Select("Select file to download:", generic).(fmt.Stringer).String()
+
+	// Reset readers so we can scan the tar file
+	// again to get the correct file reader
+	br.Seek(0, io.SeekStart)
+	xr, err = xz.NewReader(br, 0)
+	if err != nil {
+		return "", nil, err
+	}
+	tr = tar.NewReader(xr)
+
+	var fr io.Reader
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return "", nil, err
+		}
+
+		if header.Name == selectedFile {
+			fr = tr
+			break
+		}
+	}
+	// return base of selected file since tar
+	// files usually have folders inside
+	return filepath.Base(selectedFile), fr, nil
+}
+
+func ProcessURL(gf *FilteredAsset) (string, io.Reader, int64, error) {
+	// We're not closing the body here since the caller is in charge of that
+	res, err := http.Get(gf.URL)
+	log.Debugf("Checking binary from %s", gf.URL)
+	if err != nil {
+		return "", nil, 0, err
+	}
+
+	if res.StatusCode > 299 || res.StatusCode < 200 {
+		return "", nil, 0, fmt.Errorf("%d response when checking binary from %s", res.StatusCode, gf.URL)
+	}
+
+	name, outputFile, err := ProcessReader(gf.Name, res.Body)
+	if err != nil {
+		return "", nil, 0, err
+	}
+	return name, outputFile, res.ContentLength, err
+}
+
+func ProcessReader(name string, r io.Reader) (string, io.Reader, error) {
+	var buf bytes.Buffer
+	tee := io.TeeReader(r, &buf)
+
+	t, err := filetype.MatchReader(tee)
+	if err != nil {
+		return "", nil, err
+	}
+
+	var outputFile = io.MultiReader(&buf, r)
+
+	type processorFunc func(r io.Reader) (string, io.Reader, error)
+	var processor processorFunc
+	switch t {
+	case matchers.TypeZip:
+		processor = ProcessZip
+	case matchers.TypeGz:
+		processor = ProcessTarGz
+	case matchers.TypeXz:
+		processor = ProcessTarXz
+	}
+	if processor != nil {
```

Comment:
The idea here is that if the Reader is an archive we recursively check the result (could help split up .tar and .gz/.xz handling in a future commit), and if it's not an archive we return the result.

Reply:
I've separated out tar (unarchiving) processing from decompressing (gz, xz, ...) for reuse and decoupling purposes. I've also marked all but assets.ProcessURL as private functions.

Reply:
Something that would be nice to achieve is a proper progess bar at The Right Time(tm), i.e. during decompressing or unarchiving, but that's probably not straightforward. Maybe they could be done as separate progress bars, but then again they can happen simultaneously (unarchiving a tar file through the gunzip reader)...

Reply:
Won't fix this in this PR, maybe next time :)